source/common/stats/thread_local_store.cc

```diff
@@ -156,6 +156,26 @@ void ThreadLocalStoreImpl::clearScopeFromCaches(uint64_t scope_id) {
   }
 }
 
+absl::string_view ThreadLocalStoreImpl::truncateStatNameIfNeeded(absl::string_view name) {
+  // If the main allocator requires stat name truncation, warn and truncate, before
+  // attempting to allocate.
+  if (alloc_.requiresBoundedStatNameSize()) {
```

Comment:
nit: as I mentioned in person, I think the encapsulation would be cleaner if the allocators were constructed with their truncation settings and then they could handle any necessary truncation themselves (the fallback heap allocator would be set up to truncate, while the non-hot restart heap allocator would be set up to not truncate).  Here, it seems like we're half delegating that responsibility by giving them a method to communicate whether they require truncation, but not giving them the responsibility to read the settings and do that truncation within `alloc()`.

However, I think we should just try to get this PR in as is rather than making more interface changes, so this seems fine for now.

Reply:
Yeah I think that's a follow-up change.